ide/projectapi/src/org/netbeans/api/project/ProjectActionContext.java

```diff
@@ -0,0 +1,402 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.api.project;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import org.netbeans.api.annotations.common.CheckForNull;
+import org.netbeans.api.annotations.common.NonNull;
+import org.netbeans.api.project.Project;
+import org.netbeans.spi.project.ProjectConfiguration;
+import org.openide.util.Lookup;
+import org.openide.util.lookup.Lookups;
+import org.openide.util.lookup.ProxyLookup;
+
+/**
+ * This represents context for a project model query. The context describe a project action
+ * for which the query should be evaluated, project configuration (if not the active one), 
+ * and possible custom overrides.
+ * {@link ProjectConfiguration} or the IDE-supported project action and the associated properties, profiles, or Lookup may
+ * affect behaviour of queries across the IDE.
+ * <p>
+ * Properties map to user-settable properties of the build system, such as system properties in Maven
+ * or project properties in Gradle.
+ * <p>
+ * Profiles map to profiles in Maven, but do not have any effect in Gradle at the moment; this might change.
+ * <p>
+ * Instances of {@link ProjectActionContext} may be passed to some queries as (part of the) parameters explicitly. In case
+ * the API does not support such flexibility directly, the implementation(s) may - the context may be published for the
+ * query computation using {@link #apply}; the implementation may then use {@link #find} to obtain a {@link ProjectActionContext}
+ * effective for that project.
+ * <p>
+ * <b>Important note:</b> Not all project queries support {@link ProjectActionContext}. Queries which do should mention that
+ * fact in their documentation, or expose it in their APIs.
+ * 
+ * @author sdedic
+ * @since 1.89
+ */
+public final class ProjectActionContext {
+    private final Project project;
+
+    ProjectActionContext(Project project) {
+        this.project = project;
+    }
+   
+    /**
+     * The project action.
+     */
+    private String projectAction;
+    
+    /**
+     * The desired project configuration
+     */
+    private ProjectConfiguration configuration;
+    private Map<String, String> properties;
+    private Set<String> profiles;
+    private Lookup lookup;
+
+    /**
+     * Returns the project this context applies to. Having {@link Project} as a property allows multiple
+     * contexts, individually for each project, to be active.
+     * @return the target project.
+     */
+    public @NonNull Project getProject() {
```

Comment:
There are many possible null returns in this class. That should be avoided unless it brings a meaning.

Reply:
this particular one does not ;))  the others were addressed in a1c2f1b8c6b2